### PR TITLE
feat(sys): High-level personality(2) wrappers

### LINKED ISSUE #1330: Support for `personality(2)`?
Hi there! Thanks a ton for these bindings, they're a pleasure to use.

Would the nix team be interested in support for the (Linux-specific) [`personality(2)`](https://man7.org/linux/man-pages/man2/personality.2.html) syscall? It's currently supported in the `libc` crate but doesn't have higher level bindings to turn the `persona` argument into a `bitflags!` enum or to map the return value into a `Result`.

It's something I'm interested in (and capable of) implementing, if desired.

### PR DESCRIPTION
**Pull request by woodruffw**
Adds a high level `Persona` bitflags enum, as well as `personality::get()` and `personality::set()` for interacting with `libc::personality()`.

Closes #1330.

See also https://github.com/rust-lang/libc/pull/1974.

### CODE REVIEW

File: src/sys/personality.rs
Comment (woodruffw) [inline src/sys/personality.rs:8]: N.B.: These casts won't be necessary if/when https://github.com/rust-lang/libc/pull/1974 lands.

**Review by asomers — CHANGES_REQUESTED** [review]
I think this would be worthwhile.  It just needs a few things:
* To be rebased once the libc PR lands
* tests (doc tests are ok)
* CHANGELOG entry

**Comment by woodruffw** [comment]
Sounds good; I can do all of those.

**Comment by woodruffw** [comment]
Okay, I've added some doctests and a CHANGELOG entry. I'll also rebase once that PR goes through.

**Comment by asomers** [comment]
This looks good.  You'll also need to rebase to pick up a Nix fix for those libc deprecations.

**Comment by woodruffw** [comment]
Looks like musl libc doesn't support some of these persona values, so I went ahead and `cfg`'d them out.

Upsteam musl change: https://www.openwall.com/lists/musl/2020/11/18/1

**Comment by woodruffw** [comment]
Last CI failure looks unrelated:

```
test test_drop ... error: process didn't exit successfully: `/linux-runner x86_64 /target/x86_64-unknown-linux-gnu/release/deps/test_aio_drop-d7c4918b8a06891c` (signal: 1, SIGHUP: hangup)
```

**Comment by asomers** [comment]
Please rebase to fix the CHANGELOG.

**Comment by woodruffw** [comment]
I've updated this PR to not be blocked on libc: the signature of `personality(2)` is `c_ulong -> c_int` where `c_int` is always guaranteed to hold a valid persona, so `c_int` is the correct underlying type.

**Comment by asomers** [comment]
> I've updated this PR to not be blocked on libc: the signature of `personality(2)` is `c_ulong -> c_int` where `c_int` is always guaranteed to hold a valid persona, so `c_int` is the correct underlying type.

lol.  The C API doesn't make much sense, then.

**Comment by asomers** [comment]
This looks good now.  It just needs a squash.  And you'll probably want to rebase at the same time, or you may get blocked waiting for Travis.

**Comment by woodruffw** [comment]
> lol. The C API doesn't make much sense, then.

Yep. My educated guess is that it's `unsigned long` because every syscall parameter is (or was? it's been a moment since I've checked) `unsigned long` at the end of the day, and whoever wrapped it in glibc et al. got lazy. Oh well.



> This looks good now. It just needs a squash. And you'll probably want to rebase at the same time, or you may get blocked waiting for Travis.

:+1: I'll squash and rebase tomorrow.

**Comment by woodruffw** [comment]
Okay, squashed and rebased. Let me know if there's anything else (&c &c)

**Comment by woodruffw** [comment]
Unrelated failure?

```
failures:

---- src/unistd.rs - unistd::alarm (line 1636) stdout ----
thread 'src/unistd.rs - unistd::alarm (line 1636)' panicked at 'test executable failed:

thread 'main' panicked at 'assertion failed: start.elapsed() >= Duration::from_secs(1)', src/unistd.rs:21:1
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

', src/librustdoc/test.rs:341:17
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by asomers** [comment]
Try rebasing again.

**Comment by woodruffw** [comment]
Rebased!

File: CHANGELOG.md
Comment (asomers) [inline CHANGELOG.md:11]: As in the other PR, `### Changed` got duped.

**Review by asomers — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (woodruffw) [reply CHANGELOG.md]: Whoops. Fixed here and there.

**Comment by woodruffw** [comment]
(Same unrelated failure as before).

**Review by asomers — APPROVED** [review]
bors r+

**Pull request merged**
